src/Renci.SshNet/Security/Cryptography/DsaKey.cs

```diff
@@ -108,31 +118,6 @@ public DsaKey(SshKeyData publicKeyData)
             DSA = LoadDSA();
         }
 
-        /// <summary>
-        /// Initializes a new instance of the <see cref="DsaKey"/> class.
-        /// </summary>
-        /// <param name="privateKeyData">DER encoded private key data.</param>
-        public DsaKey(byte[] privateKeyData)
```

Comment:
Is it necessary to remove this?

Same applies to the other key constructors

Reply:
I remove it because there could be other DER data structures. The current Key constructors with byte array parameter assume the data is DER encoded _OpenSSL traditional format_ which is not always true. It could be PKCS#8 format as well.

In practice, most package consumers are not necessarily call Key constructors directly. It might not bring huge impacts. 

Anyway, to gracefully deprecate the constructors with byte array parameter, we can apply `Obsolete` attribute to the constructors and remove them in the future release. What do you think? @Rob-Hague

Reply:
I am sorry, I do not think there is a reason to break the existing contract.

I agree there is probably not many people using these constructors, and if we were starting from scratch we might do it differently. And although we make breaking changes practically every release, I think we can justify each of them. I do not believe this one can be justified.

As library developers, when we make a breaking change we are telling the user: "we expect you to bear the pain for our work". As a user, I can sometimes accept that when the user experience is improved and when it is easy to react to the change. Otherwise I find it annoying and frankly inconsiderate from the developers.

Whether we make this change or not, we experience no pain, so we should not cause any pain to users. We can instead improve things here by clarifying in the documentation that this constructor expects an OpenSSL format.

Reply:
I understand the situation, I just don't think it is worth deleting the constructor or changing its behaviour, unless you can somehow detect which format it is using while reading. If not, then the constructor should stay the same

Reply:
Then how about keep the constructor but apply the `Obsolete` attribute? Looking at the industry, the PKCS#1 is being replaced by PKCS#8.

Reply:
This improvement does not require changing public API. It brings us zero benefits, and brings detriment to users

Reply:
If everyone agrees, then I can add them back. Would you mind sharing your thoughts here? @drieseng @WojciechNagorski @mus65

Reply:
I agree removing it is too much. But imho it should at least be obsoleted since it is ambigious and confusing (.NET 9 recently [made a similiar API obsolete](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview7/libraries.md#changes-to-x509-certificate-loading)). But then there should be an alternative public API people can move to (I assume there is? I haven't checked the whole PR).

Reply:
> I agree removing it is too much. But imho it should at least be obsoleted since it is ambigious and confusing (.NET 9 recently [made a similiar API obsolete](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview7/libraries.md#changes-to-x509-certificate-loading)). But then there should be an alternative public API people can move to (I assume there is? I haven't checked the whole PR).

Thanks for sharing the info. I know I'm not alone.

BTW, Microsoft calls their new design **"one method, one purpose"** which is exactly what I want to apply for method `static string GetCurveOid(string curve_s)`.